### ISSUE TITLE
False negative about UnusedAssignment

### ISSUE THREAD

**Issue by Zustin** [pmd/pmd#3975]
**Affects PMD Version:6.42.0** 

**Rule:UnusedAssignment**

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment>

**Description:**
I found an inconsistent outcomes between equivalent code examples below. `method1` and `method2` are equivalent methods. However, PMD report two warnings in `method1`, and only one warning in `method2`.

**Code Sample demonstrating the issue:**
```java
boolean get() {
    return true;
}
void method1() {
    boolean halfway = false;  // report a warning here
    try {
        halfway = true; // report a warning here
    } catch(Exception e) {
        System.out.println(halfway);
    }
}
void method2() {
    boolean halfway = false;  // report a warning here
    try {
        halfway = get(); 
    } catch(Exception e) {
        System.out.println(halfway);
    }
}
```

**Expected outcome: Two warnings**

PMD should report a violation at line 15, but doesn't. This is a false-negative.

**Running PMD through:** *[CLI]*

**Comment by oowekyala** [comment]
Thanks for the report.
Method calls are considered potentially throwing exceptions, which explains the difference (both methods are not equivalent).
- in method 1, nothing can throw in the `try` block, so the catch block is effectively unreachable.  Accordingly, neither assignment is used so both warnings are legit
- in method 2, 2 execution paths are possible:
  - either `get()`  throws, in which case the second assignment is not executed, and the first is used in the catch block,
  - or `get()` does not throw. The first assignment is overwritten, and the second is never used because the catch block is not reached, so neither assignment is used.

I think in method2, we should actually report only 1 warning, but on the second assignment, as that one is actually never used, while the first might be if `get` throws. So this is not really a FN but more a bug/inconsistent behaviour.

**Comment by jsotuyod** [comment]
PMD 7.0.0 will report on lines 6, 8 and 16; but not on 14. This seems appropriate.
